### ISSUE TITLE
The diagnostics do not take into account unsaved changes

### ISSUE THREAD

**Issue by Shatur95** [clangd/clangd#488]
**Steps to reproduce**

1. Open any class
2. Create a member function for it with any name / parameters
3. Open .cpp file
4. Create any **correct** implementation of the previously created function.

**Expected behavior**

Diagnostics say that everything is okay.

**Current behavior**

Diagnostics say that such function does not exits until saving header.

**System information**

Output of `clangd --version`: 10.0.1
clangd extension version: v0.1.6
Operating system: Arch Linux

**Notes**

It works correctly in `QtCreator` that use `clangd` for diagnostics.

**Comment by Shatur95** [comment]
Thanks for the explanation. Sorry, I thought the problem was in the extension, because QtCreator uses `clangd` but respects modified files (maybe by some kind of hack).
Is it possible to add an option to take into account all changes in the editor instead only for the current one? VSCode (C++ extension by Microsoft), QtCreator and CLion display diagnostics for the unchanged files. This is because when you work in the IDE, you are almost always modifying files within it.

**Comment by kadircet** [comment]
no unfortunately there isn't such an option :(

**Comment by Shatur95** [comment]
Yes, I know there is no such command line option now, but what if implement it in future? This would allow external programs / plugins to use it instead of manually implementing this function. I faced the same problem with `clangd` with Vim.
So, if you like my proposal - please, leave this issue open.

**Comment by njames93** [comment]
Just a quick drive by, according to LSP spec. All files open in the editor should be treated as the truth, overriding what's stored on disk. In this regard clangd isn't following the spec.
Maybe clangd should be tweaked to always use the open files when including headers, with the option to revert to its current behaviour. Given both approaches have their own advantages and disadvantages, this method would likely please most.

**Comment by njames93** [comment]
@kadircet Maybe I'm thinking too much into this. But how would a vfs that uses the most recently updated file when adding the header includes. This would require storing modification times with the file contents on the server and comparing that to the on disk time and using the newer of the 2. 

This would handle cases where `git pull` updates a file you currently have open even if the editor doesn't respond to that change, while improving the support for working on a header/source file interchangeably without having to save in between.

It an editor does see the external file change and reports it to the server, there is still no issue as both file systems will contain the same date regardless of differences in modify time. 

In any case this behaviour should also be behind the same command line flag. 
```
enum HeaderMode{
  OnDisk, // Current behaviour. 
  InEditor, // What LSP says should happen, should really be the default. 
  MostRecent // What I'm proposing here. 
};```

I feel like this approach, while maybe a bit of a pain to implement, will result in a far better user experience.

**Comment by VolodymyrZibarov** [comment]
> > This is very inconvenient behavior and I do not see any case when it could be useful.
> 
> clangd tries to reflect the state of the real build as much as possible. hence we always make use of the file state on disk rather than your editor, the only exception is the main file and it is deliberately done to provide you with diagnostics as you type and all the other features (hover, go-to-def, refs, etc.) for the code you've just written.
> 
I would argue about "the only exception is the main file", as many editors is able to edit 2 or more files simultaneously, so there is no "main" file. There is all files currently opened, I wouldn't distinguish between them. I mean, you can have an editor splitted into 2 halves with 2 different files. You quickly switching between them with mouse, so for user this is the same as editing one file, because he sees it continuously and work with both of them. Going forward with this, I would say that the way user work with files in editor (ctrl-tabbing between them, or opened 2 or more together) shouldn't affect the behavior of clangd. You have all opened files, they all "main".

**Comment by sam-mccall** [comment]
~All LSP operations target a file, and diagnostics are shown in a particular file. That's what "main file" refers to, and it's perfectly well-defined.

@njames93 you added much of the infra for this, and it is really useful for refactoring. I forget at what point we got stuck, but should we finish that and make it a hidden option (command line flag) to let people test it?

I suspect it's going to be slow on large projects but if I pushed back on trying it (which is my recollection) that was probably wrong.

**Comment by njames93** [comment]
I'll have another look, been pretty busy over the last few months but I've got time now

**Comment by njames93** [comment]
If you build a trunk version of clangd now that patch has landed, it now accepts the command line flag `use-dirty-headers` that should have this functionality. If you do use it and have any feedback that'd be greatly appreciated. It will likely be clangd-15 before this actually becomes an officially supported feature.

**Comment by voertler** [comment]
I think this issue can be closed as --use-dirty-headers  is now part of clangd

**Comment by HighCommander4** [comment]
> I think this issue can be closed as --use-dirty-headers is now part of clangd

I think the remaining open question is whether we should consider making this the default.
